### ISSUE TITLE
"Error: The file “nvram.bin” doesn’t exist." when pushing an uninstalled VM

### ISSUE THREAD

**Issue by andrewhamon** [cirruslabs/tart#860]
I tried making a tart VM where I haven't even gone through the installer. The reason for this is to improve the iteration speed of boot commands - installing a VM from an ipsw is a rather slow operation, but booting to the installer on an already-created tart VM is very fast.

Full error output:

```
pushing discord-macos-installer to gcr.io/discord-development/discord-macos-installer:latest...
pushing config...
pushing disk... this will take a while...
100%
pushing NVRAM...
Error: The file “nvram.bin” doesn’t exist.
```

Tart version:

```
tart --version 
2.13.0
```

**Comment by fkorotkov** [comment]
What are the steps to reproduce? You created a VM from an IPSW but didn't go through the "Hello" installation?

**Comment by andrewhamon** [comment]
Correct. Here is my packer template:

```hcl
// This builds a macOS image which boots to the macOS installer. Keeping this as
// an independent artifact allows us to iterate on the install process without
// incurring the cost of unpacking the ipsw on each run (which can take several
// minutes).
//
// Building this is slightly jank - once the VM boots, packer will wait
// indefinitely for it to shut down. So we have to get through just enough of
// the installer to trigger a shutdown.

variable "ipsw-url" {
  type =  string
  default = "https://updates.cdn-apple.com/2024SpringFCS/fullrestores/062-01897/C874907B-9F82-4109-87EB-6B3C9BF1507D/UniversalMac_14.5_23F79_Restore.ipsw"
}

source "tart-cli" "macos-ipsw" {
  from_ipsw    = var.ipsw-url
  vm_name      = "discord-macos-installer"
  disk_size_gb = 40
  headless = true
  // Don't try to establish an SSH connection to the VM
  communicator = "none"

  // Allow sufficient time for IPSW install
  create_grace_time = "30s"

  boot_command = [
    # Wait for the VM to boot
    "<wait60s>",

    # Launch the installer
    "<spacebar>",
    "<wait2s>",

    // Select English - can't shut down until we get past this screen
    "<enter>",
    "<wait15s>",

    // Shut down
    "<leftAltOn>q",
    "<wait2s>",
    "<tab><spacebar>",
  ]
}

build {
  sources = ["source.tart-cli.macos-ipsw"]
}

packer {
  required_plugins {
    tart = {
      version = ">= 1.12.0"
      source  = "github.com/cirruslabs/tart"
    }
  }
}
```

**Comment by andrewhamon** [comment]
Hmm... I can not seem to reproduce this. No idea what happened.

**Comment by fkorotkov** [comment]
Yea, it should get created here:

https://github.com/cirruslabs/tart/blob/679289d7ab1c7a5bfd4354fac2ca9ad081e2e88c/Sources/tart/VM.swift#L171-L172

This method creates the files based on the [docs](https://developer.apple.com/documentation/virtualization/vzmacauxiliarystorage/3816042-init).

Did you run any commands in parallel while the image was getting pushed?

**Comment by fkorotkov** [comment]
I think I know what happened. Tart has a "prunning" mechanism that automatically cleans disk space when you clone a new image. It can be disabled though, please check `tart clone --help` and look for `TART_NO_AUTO_PRUNE`.

So I guess you run a command while the push was happening and the image got removed. Proposing a fix in #861 but will delegate to @edigaryev for the final word since he implemented the prunning,

**Comment by andrewhamon** [comment]
> Did you run any commands in parallel while the image was getting pushed?

Yes, definitely possible and plausible that I ran other tart commands during the (lengthy) push.

I think the most plausible explanation is that I started and then canceled a packer build which deleted the image of the same name (the packer plugin seems to delete the vm if it already exists).

I've got plenty of space on my SSD, so its not clear to me that `TART_NO_AUTO_PRUNE` would be to blame.

I also don't think your proposed fix would help much here since the VM I was pushing was not running.

I wonder if a more robust solution might be to clone a VM before pushing, to ensure the VM is unlikely to be tampered with in the middle of the push.
